subsys/lw_sched/lw_scheduler.c

```diff
@@ -0,0 +1,179 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Copyright(c) 2023 Intel Corporation
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/lw_sched/lw_sched.h>
+
+extern void lw_task_insert(struct lw_scheduler *sched, struct lw_task *task);
+
+static void timer_expiry(struct k_timer *timer)
+{
+	struct lw_scheduler *sched;
+
+	/* Get the scheduler based on the timer */
+
+	sched = CONTAINER_OF(timer, struct lw_scheduler, timer);
+
+	k_sem_give(&sched->sem);   /* Wake the scheduler's thread */
+}
+
+static void lw_scheduler_reorder(struct lw_scheduler *sched)
+{
+	sys_dnode_t *reorder;
+	struct lw_task *task;
+
+	/*
+	 * Loop through each node in the reorder list. Remove the associated
+	 * task from the list of tasks and insert it at the correct priority.
+	 */
+
+	while ((reorder = sys_dlist_get(&sched->reorder)) != NULL) {
+		task = CONTAINER_OF(reorder, struct lw_task, reorder);
+		if (task->state != LW_TASK_ABORT) {
+			if (sys_dnode_is_linked(&task->node)) {
+				sys_dlist_remove(&task->node);
+			}
+			task->order = task->new_order;
+			lw_task_insert(sched, task);
+		}
+	}
+}
+
+static void lw_scheduler_entry(void *p1, void *p2, void *p3)
+{
+	struct lw_scheduler *sched = p1;
+	struct lw_task *task;
+	sys_dnode_t *node;
+	uint32_t  state;
+	k_spinlock_key_t  key;
+	sys_dnode_t  dummy;
+
+	while (1) {
+
+		/* Wait until the timer gives the semaphore */
+
+		k_sem_take(&sched->sem, K_FOREVER);
```

Comment:
I think you could ditch the semaphore and synchronize the thread to the timer directly with `k_timer_status_sync`.